### ISSUE TITLE
Probably improper signature verification

### ISSUE THREAD

**Issue by Dartt0n** [aiogram/aiogram#1545]
# tl;dr
https://github.com/aiogram/aiogram/blob/a3d6c1615ee07eb50577b6a1b6323a8657fd9f20/aiogram/utils/web_app.py#L133
Shouldn't it be `secret_key = hashlib.sha256(token.encode())`?

# details
I'm trying to validate data received in query parameters when webapp is opened. However, current signature verification resulted in hash mismatch. After some research, I found that JS library `grammy` uses different approach for signature verification, which works for me. Question is, is aiogram implementation is incorrect, or I'm doing something wrong? When I replace the referenced line with `secret_key = hashlib.sha256(token.encode())` everything works fine for me.

Reference grammy implementation: https://github.com/grammyjs/validator/blob/10464f81129659c8392d936103da54fbd22ee4f4/src/mod.ts#L8

**Comment by Dartt0n** [comment]
Code snippet which works for me
```python
payload = request.query # request is aiohttp.web.Request
...

init_data = sorted(parse_qsl(urlencode(payload)))
data_check_string = "\n".join(f"{k}={v}" for k, v in init_data if k != "hash")
hash_ = data["hash"]

secret_key = sha256(telegram_token.encode())

calculated_hash = hmac.new(
    key=secret_key.digest(), msg=data_check_string.encode(), digestmod=sha256
).hexdigest()

return calculated_hash == hash_
```

**Comment by JrooTJunior** [comment]
This validation is correct. Just try to use it.

I test it with this code and it works as expected:

<details>
<summary>Send web app</summary>

```python
WEBAPP_URL = "https://<code>.ngrok-free.app"

@example_router.message(Command("webapp"))
async def command_webapp(message: Message, bot: Bot):
    await message.answer(
        "Webapp",
        reply_markup=InlineKeyboardMarkup(
            inline_keyboard=[
                [
                    InlineKeyboardButton(
                        text="webapp",
                        web_app=WebAppInfo(url=WEBAPP_URL),
                    )
                ]
            ]
        ),
    )
```
</details>


<details>
<summary>Web server</summary>

```python
from os import getenv

from aiohttp import web

from aiogram.utils.web_app import safe_parse_webapp_init_data

TOKEN = getenv("TELEGRAM_TOKEN")


INDEX = """<!DOCTYPE html>
<html>
<head>
    <title>aiogram</title>
    <script src="https://telegram.org/js/telegram-web-app.js"></script>
</head>
<body>
<h1>Profile</h1>
<div id="profile">loading...</div>

<script type="text/javascript">
    checkUser(window.Telegram.WebApp.initData);
    async function checkUser(initData) {
        const response = await fetch("/check", {
            method: "POST",
            body: JSON.stringify({
                "data": initData,
            }),
            headers: {
                "Content-Type": "application/json"
            }
        })
        if (response.status !== 200) {
            document.getElementById("profile").innerText = "unauthorized"
            return
        }
        let data = await response.json()

        let pre = document.createElement("pre")
        pre.innerText = JSON.stringify(data, null, 2)
        document.getElementById("profile").replaceChild(pre, document.getElementById("profile").firstChild)
    }
</script>
</body>
</html>
"""


async def index(request: web.Request):
    return web.Response(text=INDEX, content_type="text/html")


async def check_user(request: web.Request):
    payload = await request.json()
    user = safe_parse_webapp_init_data(TOKEN, payload["data"])
    return web.json_response(user.model_dump(mode="json"))


def main():
    app = web.Application()
    app.router.add_get("/", index)
    app.router.add_post("/check", check_user)

    web.run_app(app, port=8075)


if __name__ == "__main__":
    main()
```
</details>

<details>
<summary>Result</summary>

![image](https://github.com/user-attachments/assets/99414aee-3e97-4316-8f89-56545c43573d)
</details>

**Comment by Dartt0n** [comment]
Recently, I realized that Telegram uses a similar but slightly different approaches to generating hashes for messages. This one is correct for Telegram web apps, which are open from the bots. However, for the Telegram login widget, the correct approach is to use just the SHA256 hash of the token. This is weird, since these two requests are exactly the same. The only difference is in the hashes. In our case, we have ended up implementing two different routes: one for opening the web app inside Telegram (and verifying it using AIogram utilities) and one for using the Telegram login widget (with hash verification, as described in the issue description).

As a part related to alogram, this issue may be closed as irrelevant, since the problem lies in the design and documentation of telegram services, but may be additional utility could be added to the aiogram library to enhance the developer experience.

**Comment by JrooTJunior** [comment]
It would be nice to have such a utility inside the framework, feel free to contribute

**Comment by Dartt0n** [comment]
I returned to work and found that this feature had already been implemented, but under a slightly different name:
https://github.com/aiogram/aiogram/blob/a3d6c1615ee07eb50577b6a1b6323a8657fd9f20/aiogram/utils/auth_widget.py#L6
